merkle-tree/indexed/tests/tests.rs

```diff
@@ -110,63 +113,281 @@ fn relayer_update<H>(
             lowest_from_queue.index as u16,
-            nullifier,
+            nullifier.index as u16,
+            nullifier.next_index as u16,
+            nullifier.next_value,
             old_low_nullifier,
             low_nullifier_proof,
-        );
+        ) {
+            update_errors.push(e);
+        }
 
         // Update off-chain tree.
         relayer_merkle_tree
-            .update(
-                new_low_nullifier,
-                new_low_nullifier.index,
-                nullifier,
-                nullifier.index,
-            )
+            .update(new_low_nullifier, nullifier)
             .unwrap();
 
         // Insert the element to the indexing array.
         relayer_indexing_array
-            .append_with_low_element_index(new_low_nullifier.index, nullifier.value);
+            .append_with_low_element_index(new_low_nullifier.index, nullifier.value)
+            .unwrap();
+    }
+
+    if update_errors.is_empty() {
+        Ok(())
+    } else {
+        Err(RelayerUpdateError::MerkleTreeUpdate(update_errors))
     }
 }
 
-fn cpi_update<H>(
-    // PDAs
-    queue: &mut RefMut<'_, IndexingArray<H, QUEUE_ELEMENTS>>,
-    merkle_tree: &mut RefMut<
-        '_,
+/// Tests the valid case of:
+///
+/// * Inserting nullifiers to the queue.
+/// * Calling the relayer to update the on-chain nullifier Merkle tree.
+fn insert_and_update<H>()
+where
+    H: Hasher,
+{
+    // On-chain PDAs.
+    let onchain_queue: RefCell<IndexingArray<H, QUEUE_ELEMENTS>> =
+        RefCell::new(IndexingArray::default());
+    let onchain_tree: RefCell<
         IndexedMerkleTree<H, MERKLE_TREE_HEIGHT, MERKLE_TREE_CHANGELOG, MERKLE_TREE_ROOTS>,
-    >,
-    // Instruction data
-    changelog_index: u16,
-    queue_index: u16,
-    new_nullifier: IndexingElement,
-    old_low_nullifier: IndexingElement,
-    low_nullifier_proof: [[u8; 32]; MERKLE_TREE_HEIGHT],
-) where
+    > = RefCell::new(IndexedMerkleTree::default());
+    onchain_tree.borrow_mut().init().unwrap();
+
+    // Insert a pair of nullifiers.
+    let nullifier1: [u8; 32] = [
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 30,
+    ];
+    let nullifier2: [u8; 32] = [
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 10,
+    ];
+    program_insert::<H>(onchain_queue.borrow_mut(), [nullifier1, nullifier2]).unwrap();
+
+    // Insert an another pair of nullifiers.
+    let nullifier3: [u8; 32] = [
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 20,
+    ];
+    let nullifier4: [u8; 32] = [
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 50,
+    ];
+    program_insert::<H>(onchain_queue.borrow_mut(), [nullifier3, nullifier4]).unwrap();
+
+    // Call relayer to update the tree.
+    relayer_update::<H>(
```

Comment:
Are there asserts that the update was successful, not only that the update function didn't throw any errors?

Reply:
You could add asserts at the end of relayer_update so that these are always triggered once the function runs.

Reply:
Addressed here https://github.com/Lightprotocol/light-protocol/pull/428/commits/27bca92541ae25d10988e3fe7ba3b304ab1811b0

Reply:
thanks!